File: 1127-last-stone-weight/1127-last-stone-weight.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function(stones) {
    const maxHeap = [];

    /**
     * Добавляет элемент в кучу и восстанавливает свойство кучи.
     * @param {number} val - Значение для вставки.
     */
    function insert(val) {
        maxHeap.push(val);
        bubbleUp(maxHeap.length - 1);
    }

    /**
     * Извлекает максимальный элемент из кучи.
     * @returns {number} - Максимальный элемент.
     */
    function extractMax() {
        if (maxHeap.length === 0) return null;
        if (maxHeap.length === 1) return maxHeap.pop();

        const max = maxHeap[0];
        maxHeap[0] = maxHeap.pop(); 
        bubbleDown(0);
        return max;
    }

    /**
     * Возвращает максимальный элемент без удаления.
     * @returns {number}
     */
    function peek() {
        return maxHeap[0];
    }

    /**
     * Перемещает элемент вверх по куче, пока свойство макс-кучи не выполнено.
     * @param {number} index - Индекс элемента.
     */
    function bubbleUp(index) {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (maxHeap[parentIndex] < maxHeap[index]) {
                swap(parentIndex, index);
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    /**
     * Перемещает элемент вниз по куче, пока свойство макс-кучи не выполнено.
     * @param {number} index - Индекс элемента.
     */
    function bubbleDown(index) {
        const lastIndex = maxHeap.length - 1;
        while (true) {
            const leftChildIndex = 2 * index + 1;
            const rightChildIndex = 2 * index + 2;
            let largestIndex = index;

            if (leftChildIndex <= lastIndex && maxHeap[leftChildIndex] > maxHeap[largestIndex]) {
                largestIndex = leftChildIndex;
            }
            if (rightChildIndex <= lastIndex && maxHeap[rightChildIndex] > maxHeap[largestIndex]) {
                largestIndex = rightChildIndex;
            }

            if (largestIndex !== index) {
                swap(index, largestIndex);
                index = largestIndex;
            } else {
                break;
            }
        }
    }

    /**
     * Меняет местами два элемента в куче.
     * @param {number} i - Индекс первого элемента.
     * @param {number} j - Индекс второго элемента.
     */
    function swap(i, j) {
        [maxHeap[i], maxHeap[j]] = [maxHeap[j], maxHeap[i]];
    }

    for (const stone of stones) {
        insert(stone);
    }

    // Сталкиваем камни, пока не останется 0 или 1
    while (maxHeap.length > 1) {
        const y = extractMax(); 
        const x = extractMax(); 

        if (x !== y) {
            insert(y - x); 
        }
        // если x === y — оба уничтожаются, ничего не добавляем
    }

    return maxHeap.length === 0 ? 0 : peek();
};

File: 1127-last-stone-weight/1127-last-stone-weight.java
public class Solution {
    public int lastStoneWeight(int[] stones) {
        MaxHeap maxHeap = new MaxHeap();

        for (int stone : stones) {
            maxHeap.insert(stone);
        }

        while (maxHeap.size() > 1) {
            int y = maxHeap.extractMax(); // самый тяжёлый
            int x = maxHeap.extractMax(); // второй по тяжести

            if (x != y) {
                maxHeap.insert(y - x);
            }
        }

        return maxHeap.size() == 0 ? 0 : maxHeap.peek();
    }
}

class MaxHeap {
    private List<Integer> heap;

    public MaxHeap() {
        this.heap = new ArrayList<>();
    }

    public int peek() {
        return heap.isEmpty() ? 0 : heap.get(0);
    }

    public int size() {
        return heap.size();
    }

    public void insert(int val) {
        heap.add(val);
        bubbleUp(heap.size() - 1);
    }

    private void bubbleUp(int index) {
        while (index > 0) {
            int parentIndex = (index - 1) / 2;
            if (heap.get(parentIndex) < heap.get(index)) {
                swap(parentIndex, index);
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    private void bubbleDown(int index) {
        int lastIndex = heap.size() - 1;
        while (true) {
            int left = 2 * index + 1;
            int right = 2 * index + 2;
            int largest = index;

            if (left <= lastIndex && heap.get(left) > heap.get(largest)) {
                largest = left;
            }
            if (right <= lastIndex && heap.get(right) > heap.get(largest)) {
                largest = right;
            }

            if (largest != index) {
                swap(index, largest);
                index = largest;
            } else {
                break;
            }
        }
    }

    public int extractMax() {
        if (heap.isEmpty()) {
            return 0; // или бросить исключение, но в контексте задачи 0 допустимо
        }
        if (heap.size() == 1) {
            return heap.remove(0);
        }

        int max = heap.get(0);
        heap.set(0, heap.remove(heap.size() - 1)); // перемещаем последний элемент в корень
        bubbleDown(0);
        return max;
    }

    private void swap(int i, int j) {
        int temp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, temp);
    }
}